### PR TITLE
(Bi)conjugate Frank-Wolfe implementations

### PR DESCRIPTION
**Pull request by sboyles**
In addition to FW/CFW/BFW there is a streamlined version of the shortest path finding that allows for warm-starting... if merging with the Alg-B implementations, may also need to review bush.c to see if its calls (in initializing bushes and calculating gap) need to be changed as well.

This code has not yet been debugged yet, but it compiles warning-free

### CODE REVIEW

**Comment by pyrito** [comment]
@sboyles  , I'm ready to merge this but I'm not sure if your code base is up to date with our master branch? I was reviewing some of the changes and it looks like your branch in networks.h overwrites some of the parallel things that we added. Could you double check this?

**Comment by ribsthakkar** [comment]
We can merge in the changes for you?

File: include/networks.h
Comment (ribsthakkar) [inline include/networks.h]: Need this for parallelism

File: include/networks.h
Comment (ribsthakkar) [inline include/networks.h]: These are the mutexes to prevent concurrent updates to links

File: src/networks.c
Comment (ribsthakkar) [inline src/networks.c]: same as above

File: src/networks.c
Comment (ribsthakkar) [inline src/networks.c]: same as above

File: include/networks.h
Comment (pyrito) [reply include/networks.h]: Yeah I'm just wondering if there are other modifications like this made, and I'm wondering if Dr. Boyles is up to date. We can prob verify this though.

File: src/networks.c
Comment (ribsthakkar) [inline src/networks.c]: using network->batchSize makes it more general to even batching the tntp networks. We changed this previously

File: src/networks.c
Comment (ribsthakkar) [inline src/networks.c]: Same as comment for why we added these in the first place

File: src/networks.c
Comment (ribsthakkar) [inline src/networks.c]: We modified this and removed the modulo because this assumes demand matrix is currently batchSize, but during the reading of the network at the start on Line 280-290 of fileio.c, it will overwrite the demand matrix for just the first class

**Review by ribsthakkar — CHANGES_REQUESTED** [review]

**Comment by sboyles** [comment]
Ah, I think I am out of date.  I *think* the only real change was to
the shortest path code, I will take a closer look when I get into the
office tomorrow morning.

Easiest thing for now is probably to reject my networks.c / networks.h
push.  I'll come up with a new name for my updated shortest path
function so it won't clash with anything else and re-commit.

-Steve

On 5/11/20, Karthik Velayutham <notifications@github.com> wrote:
> @sboyles  , I'm ready to merge this but I'm not sure if your code base is up
> to date with our master branch? I was reviewing some of the changes and it
> looks like your branch in networks.h overwrites some of the parallel things
> that we added. Could you double check this?
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/spartalab/tap-b/pull/4#issuecomment-626940996


-- 
Stephen Boyles
Associate Professor
Department of Civil, Architectural, and Environmental Engineering
The University of Texas at Austin
Ph: 512-471-3548
http://sboyles.github.io

**Comment by sboyles** [comment]
OK, I think I've merged in my new networks file with your updates.  I also came up with new names for my updated shortest path functions so nothing should break in the old code (eventually we can update the references elsewhere in the code and deprecate the old functions altogether, but I don't think that's urgent.)

Also pushed an update to my heap implementation which is faster for Dijkstra

**Comment by pyrito** [comment]
@sboyles these changes are looking good. Let us know when to merge when you are done with your pushes!

**Comment by sboyles** [comment]
Cool, I'm still doing some debugging... would it be better to wait to merge until I'm done later this afternoon, or to just do it now and commit my remaining patches to master?

**Comment by ribsthakkar** [comment]
Better to have a working version before merge

**Pull request merged**
